Remove libstdc++ dependency from NativeAOT

am11 opened the pull request:
Fixes #76655.

EgorBo commented:
nice, I'm just curious - does it lead to smaller binary size?

eerhardt commented:
>  does it lead to a smaller binary size?

I don't know about the binary size (I haven't tested it), but it leads to a smaller container size because you don't need to install libstdc++ as a dependency. Here's the "docker inspect" of a small example NativeAOT app built on alpine:

![image](https://user-images.githubusercontent.com/8291187/194336865-b18df6e5-b387-4d79-b433-ba6da7c6f03c.png)

Note that the libstdc++ dependency almost doubles the starting image size.

AaronRobinsonMSFT commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:
```diff
@@ -96,7 +96,6 @@ The .NET Foundation licenses this file to you under the MIT license.
       <LinkerArg Include="-Wl,--as-needed" Condition="'$(TargetOS)' != 'OSX'" />
       <LinkerArg Include="-Wl,-e0x0" Condition="'$(NativeLib)' == 'Shared' and '$(TargetOS)' != 'OSX'" />
       <LinkerArg Include="-pthread" Condition="'$(TargetOS)' != 'OSX'" />
-      <LinkerArg Include="-lstdc++" />
```
Will this cause a local build to fail or only the testing? I'm concerned about how easy it is to introduce a C++ dependency. By doing this we will be imposing the restriction on all aspects of native code correct? I'd assume this would have knock-on effects in the GC as well. If not, we will need some way to ensure they aren't introduced and then need to be rewritten.

am11 replied:
Full `./build.sh` or `packs` subset in particular will fail to generate native code for crossgen2 binary.

AaronRobinsonMSFT replied:
This is something we should discuss for sure. Compiling cpp files with the expectation they aren't using the C++ runtime but on what is acceptable is rather annoying. I would personally prefer if we just compiled them as C if that is what we want. However, this is a much broader conversation. I think your current pass makes sense, but we need a real policy in place that we all agree on. Give me some time, but for this particular PR I think I am fine for now as long as @jkotas and others are on board.

jkotas replied:
> I have "replace new's with malloc's" bulk change (wip) stashed if needed. :)

Could you please share it? I would like to see what it looks like.

am11 replied:
https://github.com/dotnet/runtime/compare/main...am11:runtime:feature/nativeaot/convert-new_delete-to-malloc_free

This was wip when I decided to use the shim approach and stashed it. The ctor calls after malloc+nullcheck can be converted to `initialize()` pattern in pure C.

AaronRobinsonMSFT replied:
> > I have "replace new's with malloc's" bulk change (wip) stashed if needed. :)

> Could you please share it? I would like to see what it looks like.

I'm actually far less concerned about these changes.

> The ctor calls after malloc+nullcheck can be converted to initialize() pattern in pure C.

If you updated that diff with this, I'd be fine moving forward with that PR. @Maoni0 Thoughts on making more gc code more C-ish when possible?

jkotas replied:
> I'm actually far less concerned about these changes.

Yes, I would prefer something like this too.

am11 commented on src/coreclr/nativeaot/Runtime/stdcppshim.cpp:
```diff
@@ -0,0 +1,30 @@
+
+#include "CommonTypes.h"
+
+void __cxa_pure_virtual() 
```
`__cxa_pure_virtual` (which indicates bug or unusual interaction) is typically implemented as while(true) loop or abort(), so logically we won't miss much if our implementation is preferred over that of libstdc++.

For static linkage, if there are multiple conflicting symbols  we can use `--allow-multiple-definition` (as we anyway need today to link multiple NAOT static objects: https://github.com/dotnet/runtime/issues/70277#issuecomment-1157574381).

jkotas replied:
I am worried about somebody statically linking NAOT static object and static C++ libraries together.

I am not worried about somebody linking multiple NAOT static objects together. I do not think that it is interesting scenario.

jkoritzinsky replied:
Would weak symbols be a reasonable solution here? If we make this a weak symbol, then it will only be used when no C++ libraries are statically linked with a NAOT static object.

am11 replied:
> I am worried about somebody statically linking NAOT static object and static C++ libraries together.

If two objects are providing the same symbol, then `--allow-multiple-definition` avoids the `duplicate symbols` warning.

jkotas replied:
What are the C++ constructs that end up producing `__cxa_pure_virtual` calls? Can we get rid of them instead?

am11 replied:
Both compilers (clang++ and g++) expect `__cxa_pure_virtual` symbol in the binary when we use a pure virtual function `T M() = 0;`, regardless the code violates the rule of calling PVF from ctor or dtor (which we don't).

This is what we get:
 
```
 crossgen2 -> /runtime/artifacts/bin/coreclr/Linux.x64.Debug/crossgen2/linux-x64/crossgen2.dll
  Generating native code
  /usr/bin/ld: /usr/bin/ld: DWARF error: invalid or unhandled FORM value: 0x25
  /runtime/artifacts/bin/coreclr/Linux.x64.Debug/aotsdk/libRuntime.ServerGC.a(objecthandle.cpp.o):(.data.rel.ro._ZTV14IGCHandleStore[_ZTV14IGCHandleStore]+0x10): undefined reference to `__cxa_pure_virtual'
  /usr/bin/ld: /runtime/artifacts/bin/coreclr/Linux.x64.Debug/aotsdk/libRuntime.ServerGC.a(objecthandle.cpp.o):(.data.rel.ro._ZTV14IGCHandleStore[_ZTV14IGCHandleStore]+0x18): undefined reference to `__cxa_pure_virtual'
  /usr/bin/ld: /runtime/artifacts/bin/coreclr/Linux.x64.Debug/aotsdk/libRuntime.ServerGC.a(objecthandle.cpp.o):(.data.rel.ro._ZTV14IGCHandleStore[_ZTV14IGCHandleStore]+0x20): undefined reference to `__cxa_pure_virtual'
  /usr/bin/ld: /runtime/artifacts/bin/coreclr/Linux.x64.Debug/aotsdk/libRuntime.ServerGC.a(objecthandle.cpp.o):(.data.rel.ro._ZTV14IGCHandleStore[_ZTV14IGCHandleStore]+0x28): undefined reference to `__cxa_pure_virtual'
  /usr/bin/ld: /runtime/artifacts/bin/coreclr/Linux.x64.Debug/aotsdk/libRuntime.ServerGC.a(objecthandle.cpp.o):(.data.rel.ro._ZTV14IGCHandleStore[_ZTV14IGCHandleStore]+0x30): undefined reference to `__cxa_pure_virtual'
  /usr/bin/ld: /runtime/artifacts/bin/coreclr/Linux.x64.Debug/aotsdk/libRuntime.ServerGC.a(objecthandle.cpp.o):(.data.rel.ro._ZTV14IGCHandleStore[_ZTV14IGCHandleStore]+0x38): more undefined references to `__cxa_pure_virtual' follow
```

The demangled form of `_ZTV14IGCHandleStore` is `vtable for IGCHandleStore`.

While googling about the compiler internals, I stumbled upon https://elegantinvention.com/blog/information/smaller-binary-size-with-c-on-baremetal-g/. The author was also trying to build with -nostdlib and ran into this issue due to the precense of pure virtual in their code base.

> Can we get rid of them instead?

`IGCHandleStore` is defined in the gcinterface, which is now a public contract (by means of standalone-gc's 3P implementations). Otherwise, I was going with replacing all occurrences of `IGCHandleStore` with `GCHandleStore` and deleting the class. :)

jkotas replied:
We do not support 3P gc implementations. GC interface is not a public contract. We only care about the GC contract to the point of being able swap the GC implementations for some range of builds.

What is special about IGCHandleStore? We use similar pattern for other types too (e.g. `IGCHeap`) and they are not causing any problems.

Can we define `#define ABSTRACT_IMPL` macro and use it instead of `= 0;` for abstract types? It should get rid of dependency on `__cxa_pure_virtual`.

am11 replied:
> We do not support 3P gc implementations. GC interface is not a public contract.

Cool. I found implementations in the wild inheriting from IGCHandleManager to provide their own manager in custom standalone gc. Good to know we don't consider it as a public contract.

> What is special about IGCHandleStore? We use similar pattern for other types too (e.g. IGCHeap) and they are not causing any problems.

Nothing other than `g_gcGlobalHandleStore` is first one used in the initialization?

I still don't understand what is wrong with `__cxa_pure_virtual`. Could you give an example what scenario will break? There is no difference in static linking commands which I showed here https://github.com/dotnet/runtime/issues/70277.

jkotas replied:
`--allow-redefinition` can hide serious problems, it looks very fragile. I do not want to be telling anybody to use `--allow-redefinition` because of then we will need to be spending time debugging the inscrutable problems caused by it.

am11 commented on src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp:
```diff
@@ -941,13 +949,16 @@ extern "C" UInt32_BOOL ResetEvent(HANDLE event)
+
 extern "C" uint32_t GetEnvironmentVariableA(const char * name, char * buffer, uint32_t size)
 {
-    // Using std::getenv instead of getenv since it is guaranteed to be thread safe w.r.t. other
```
Good point, and of course I missed that. I haven't figured out how this can be done atomically though. 😅

Looking at `src/native/libs/System.Native/pal_environment.c` -- the way interop layer is using it without any guards, we can change it to `strdup(getenv(name))` to avoid race conditions and remove the lock. Would it make sense?

jkotas replied:
I do not think that `strdup` would fix any interesting conditions. 

I think we should just call getenv here without any locks (basically, match `pal_environment.c`).

am11 commented on src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp:
```diff
@@ -488,6 +489,13 @@ extern "C" void PalAttachThread(void* thread)
+#ifdef TARGET_LINUX
+int __cxa_thread_atexit(void (*func)(void*), void* t, void* c)
+{
+    return __cxa_thread_atexit_impl(func, t, c);
```
`__cxa_thread_atexit_impl` is available on glibc >= 2.18, but not older versions (used by centos7...) or musl-libc, so I will dynamically load the symb (with `dlsym` and co.).

But what should be the fallback implementation? call destructor under a pthread lock or a nop?

am11 replied:
My understanding is that the per-thread atexit is not implemented in those variants of libcs, it fallsbac to libstdc++ implementation or one provided by the compiler: https://reviews.llvm.org/D21803?id=62394.

am11 commented:
> nice, I'm just curious - does it lead to a smaller binary size?

Yes and no.

`linux-x64`: yes (-15kb). crossgen2's native binary size went from `45249216` -> `45233888` bytes.
`osx-arm64`: no (+1.4kb). crossgen2's native binary size went from  `67941062` -> `67942550` bytes.
 
If we convert it to pure C, then it will get slimmer on both. Right now, compiler is baking in some code in `-nostdlib` / `-nostdlib++` mode for C++ sources but it is still bulkier than pure C (and what can be achieved with `-ffreestanding`).

am11 commented on src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp:
```diff
@@ -498,14 +500,55 @@ extern "C" void PalAttachThread(void* thread)
+    link = (Link*)malloc(sizeof(Link));
+    if (link)
+    {
+        link->Next = (Link*)pthread_getspecific(key);
```
A few related observations (repo-wide):

* When C++ compiler encounters `thread_local` keyword on linux, it expects `__cxa_thread_atexit` symbol to be present (provided by libstdc++).
* C11 has an alternative `_THREAD_local`, but msvc [does not support it yet](https://developercommunity.visualstudio.com/t/C-compiler:-_Thread_local-is-not-suppore/10020499?space=62&q=_Thread_local)
  * cross-dac uses the empty declaration in hp-libunwind on windows https://github.com/dotnet/runtime/blob/0cb25b36092dac69b7acd5acb330acaf5708686a/src/native/external/libunwind_extras/configure.cmake#L22-L25
  * mono uses declspec(thread)  in C on windows https://github.com/dotnet/runtime/blob/0cb25b36092dac69b7acd5acb330acaf5708686a/src/mono/CMakeLists.txt#L272
  * nativeaot also uses declspec(thread) on windows and __thread on unix https://github.com/dotnet/runtime/blob/0cb25b36092dac69b7acd5acb330acaf5708686a/src/coreclr/nativeaot/Runtime/CommonMacros.h#L235-L238 note that __thread (gcc/clang extension) and _THREAD_LOCAL (C11) on unix have the same semantics.
  * llvm-libunwind also uses declspec(thread) in setjump/longjump sjlj on windows https://github.com/dotnet/runtime/blob/0cb25b36092dac69b7acd5acb330acaf5708686a/src/native/external/llvm-libunwind/src/Unwind-sjlj.c#L68 but we do not compile this code in nativeaot (we only compile the alternative: Unwind-EHABI.cpp / dwarf as sjlj is typically used by mingw based toolchain which we don't support)

Since C11 and C++11 are minimal requirements, we can safely use `_THREAD_LOCAL` on unix and `declspec(thread)` on windows everywhere, in order to get rid of `__thread` (gcc extension) and `thread_local` (C++11 keyword) variants.

jkotas replied:
> we can safely use _THREAD_LOCAL on unix and declspec(thread) on windows everywhere,

This sounds good to me for the minimal native AOT runtime.

CoreCLR can stay with `thread_local` since it has heavy dependency on C++ runtime. No need to change it there.

am11 replied:
> CoreCLR can stay with `thread_local` since it has heavy dependency on C++ runtime. No need to change it there.

Sure. Btw, CoreCLR is using both, `thread_local` (more often) and `declspec(thread)` (in a few places).

am11 replied:
I have made the fallback path to be taken on Unix (both Linux and macOS). To use C11's `_Thread_local` and `tss_create` etc. we would need to convert the source to `.c` to be able to use `#include <threads.h>`, as C++11's `#include <thread>` requires std::new, which then conflicts with our shim.

I will try to give a separate PR pass to convert NativeAOT sources to C (there are quite many changes to make it happen:; new->malloc, delete->free, class->struct etc.). It will also require discussions with GC team; mainly to find out if C++ to C conversion is a viable option for their development.

jkotas replied:
> requires std::new, which then conflicts with our shim.

Do you mean no-throw new overload? Can we create our own no-throw new overload to avoid conflicts?

I think converting GC to C would be too much of a stretch at this point. I think sticking with C-with-classes is much more workable.

davidfowl replied:
Sounds like we need to port the GC to C# 😄

am11 replied:
> Can we create our own no-throw new overload to avoid conflicts?

We have no-throw overload in `stdcppshim.cpp`. If we use `#include <thread>` (or `#include <algorithm>` etc.), those headers in turn use `#include <new>` (which requires libstdc++ runtime library) and conflict with our shim.

am11 replied:
> Sounds like we need to port the GC to C# 😄

There is a hope: https://github.com/dotnet/runtime/discussions/72037 🙂

am11 commented:
Failure in `windows x64 checked NativeAOT` is https://github.com/dotnet/runtime/issues/76801 and installer one is unrelated.

jkotas commented on src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp:
```diff
@@ -481,11 +477,52 @@ EXTERN_C intptr_t RhGetCurrentThunkContext()
+    pthread_key_create(&key, CallDestructors);
+}
+
+int __cxa_thread_atexit(Destructor dtor, void* arg, void*)
```
We should not need general purpose `__cxa_thread_atexit`. This can simplified to call just the thread destructor.

am11 replied:
clang gives an error:

```
  crossgen2 -> /runtime/artifacts/bin/coreclr/Linux.x64.Debug/crossgen2/linux-x64/crossgen2.dll
  Generating native code
  /usr/bin/ld: /usr/bin/ld: DWARF error: invalid or unhandled FORM value: 0x25
  /runtime/artifacts/bin/coreclr/Linux.x64.Debug/aotsdk/libRuntime.ServerGC.a(PalRedhawkUnix.cpp.o): in function `__cxx_global_var_init':
  PalRedhawkUnix.cpp:(.text.startup+0x26): undefined reference to `__cxa_thread_atexit'
```

I did not find any way to satisfy the compiler without exporting `__cxa_thread_atexit` symbol. (Of course if we switch to pure C and use TLS in a C way, then these __cxa symbols are not needed)

jkotas replied:
`__cxa_thread_atexit` is required by `thread_local TlsDestructionMonitor` that has a destructor. The only purpose of `TlsDestructionMonitor` is to notify the runtime about thread shutdown. If you delete `TlsDestructionMonitor` and change the pthread callback to notify the runtime directly, `__cxa_thread_atexit` should not be needed.

jkotas replied:
I do not think that any of this requires switching to pure C.

am11 replied:
> `__cxa_thread_atexit` is required by `thread_local TlsDestructionMonitor` that has a destructor.

It gave me segmentation fault. The compiler is either indiscriminately calling `__cxa_thread_atexit` or there is more to `thread_local` which forces the compiler to go in the "MT mode" and it generates `__cxa_thread_atexit` call.

jkotas replied:
That does not sound right. Where did the segmentation fault came from?

am11 replied:
```
  Generating native code
  crossgen2 -> /Users/adeel/projects/runtime/artifacts/bin/coreclr/OSX.arm64.Debug/crossgen2/osx-arm64/publish/
  Emitting R2R PE file: /Users/adeel/projects/runtime/artifacts/obj/Microsoft.NETCore.App.Crossgen2/Debug/net7.0/osx-arm64/S.P.C.tmp
  /var/folders/_p/q6xw9h591kx26frkrng5bfmw0000gn/T/tmp24e908c43c4e47ec9d0bbb953c01ae34.exec.cmd: line 2: 12772 Trace/BPT trap: 5       /Users/adeel/projects/runtime/artifacts/bin/coreclr/OSX.arm64.Debug/crossgen2/osx-arm64/publish/crossgen2 /Users/adeel/projects/runtime/artifacts/bin/coreclr/OSX.arm64.Debug/IL/System.Private.CoreLib.dll --out /Users/adeel/projects/runtime/artifacts/obj/Microsoft.NETCore.App.Crossgen2/Debug/net7.0/osx-arm64/S.P.C.tmp
/Users/adeel/projects/runtime/src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj(70,5): error MSB3073: The command "/Users/adeel/projects/runtime/artifacts/bin/coreclr/OSX.arm64.Debug/crossgen2/osx-arm64/publish/crossgen2 /Users/adeel/projects/runtime/artifacts/bin/coreclr/OSX.arm64.Debug/IL/System.Private.CoreLib.dll --out /Users/adeel/projects/runtime/artifacts/obj/Microsoft.NETCore.App.Crossgen2/Debug/net7.0/osx-arm64/S.P.C.tmp" exited with code 133.
```

It failed with SIGTRAP on osx-arm64 but worked on linux-x64. I can guard the `thread_local` code for non-linux.

jkotas replied:
> I can guard the thread_local code for non-linux.

Sounds good to me. We have used pthread callbacks originally and switched to `thread_local` to fix macOS specific problems in https://github.com/dotnet/corert/pull/2531. We will need to stick with `thread_local` on macOS (or come up with some other way to fix the thread shutdown crash).

jkotas commented on src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp:
```diff
@@ -481,11 +477,52 @@ EXTERN_C intptr_t RhGetCurrentThunkContext()
+
+static void CreateKey(void)
+{
+    pthread_key_create(&key, CallDestructors);
```
This can be done as part of `PalInit()`

am11 replied:
Moved `pthread_once` call to `PalInit` (which takes this factory as an arg). It is a bit safer with `PTHREAD_ONCE_INIT`. If we are sure that `PalInit` can not run into race conditions, I can inline it and remove `PTHREAD_ONCE_INIT`.

jkotas commented on src/coreclr/nativeaot/CMakeLists.txt:
```diff
@@ -30,6 +30,8 @@ if(CLR_CMAKE_HOST_UNIX)
   add_compile_options(-fno-rtti)          # Native AOT runtime doesn't use RTTI
   add_compile_options(-fno-exceptions)    # Native AOT runtime doesn't use C++ exception handling
+  add_compile_options(-nostdlib)
+  add_compile_options(-fno-threadsafe-statics)
```
Is this going to generate buggy code with race conditions? I think we need to refactor any statics in local scopes that may depend on C++ runtime instead of passing this option.

am11 replied:
Without this flag, it requires `___cxa_guard_acquire` and `__cxa_guard_release` both on linux and osx.

> `-fno-threadsafe-statics`
> Do not emit the extra code to use the routines specified in the C++ ABI for thread-safe initialization of local statics. You can use this option to reduce code size slightly in code that doesn't need to be thread-safe.

jkotas replied:
It is not ok to depend on race conditions to be reliably caught by CI.

jkotas replied:
The code that triggers this looks like this and it will have a subtle race condition with this command line option:
```
class TypeWithConstructor
{
public:
    TypeWithConstructor();
};

TypeWithConstructor* m()
{
    static TypeWithConstructor x;
    return &x;
}
```
It needs to be changed to something like this so that the static is initialized at load time and the delayed initialization is not required:
```
class TypeWithConstructor
{
public:
    TypeWithConstructor();
};

static TypeWithConstructor x;

TypeWithConstructor* m()
{
    return &x;
}
```

jkotas commented on src/coreclr/nativeaot/CMakeLists.txt:
```diff
@@ -30,6 +30,8 @@ if(CLR_CMAKE_HOST_UNIX)
 
   add_compile_options(-fno-rtti)          # Native AOT runtime doesn't use RTTI
   add_compile_options(-fno-exceptions)    # Native AOT runtime doesn't use C++ exception handling
+  add_compile_options(-nostdlib)
```
Would `-nostdlib++` be a safer option? What is kind of problem do you see when you do not pass this at all?

am11 replied:
`-nostdlib++` was not available on older versions of clang and gcc that's why I used `-nostdlib`.

> What is kind of problem do you see when you do not pass this at all?

This line is the whole point of the PR. 🙂 or did I misunderstood the question?

jkotas replied:
I thought that they key part is removing `-lstdc++`. What are the additional things that `-nostdlib` does beyond that? 

The documentation suggests that `-nostdlib` is removing the dependency on all standard libraries, including C standard libraries and built-ins libraries. I am not sure what to expect from that.

am11 replied:
Right, we are building the archive without any C++ runtime dependency, therefore`-lstdc++` is not needed during the publishing of app; `-nostdlib` just makes it easier to figure out the dependencies on libstdc++.

In the scope of `libRuntime.WorkstationGC.a`, it leaves the external linkage for C APIs. When we compile it with `clang libRuntime.WorkstationGC.a <other-objects>` during the publishing of NativeAOT app, if we specify `-nostdlib` there, then the final binary will not have the libc linkage (but we will get tons of errors if we try to do that).

Currently the only difference before and after this PR is the absence of libstdc++ dependency in the NativeAOT published app.

jkotas commented on src/coreclr/nativeaot/Runtime/inc/CommonTypes.h:
```diff
@@ -6,9 +6,30 @@
+
+extern "C" void __cxa_pure_virtual();
+
+namespace std
```
There can be a collision with C++ library that the user may want to link statically. I think it would be best to avoid stomping on or reimplementing anything that C++ runtime library provides, prefer refactoring the code to avoid the dependencies C++ runtime library.

am11 replied:
Changed `std` to `nativeaot`.

jkotas commented on src/coreclr/nativeaot/Runtime/inc/CommonTypes.h:
```diff
@@ -6,10 +6,34 @@
+
+void* operator new(size_t n, const nativeaot::nothrow_t&) noexcept;
+void* operator new[](size_t n, const nativeaot::nothrow_t&) noexcept;
+void operator delete(void *p) noexcept;
+void operator delete[](void *p) noexcept;
```
Is it possible to make the operator delete to have a local scope somehow?

am11 replied:
I haven't spent much time but the initial attempt was failing when I put it in the namespace. (this is how we do it in `util_nodependencies.cpp`). Will try again.

am11 replied:
Unfortunately operators are not allowed in namespaces. It is possible to declare them in classes, but we can't use them with `using` declaration; we get `error: using declaration cannot refer to class member`.

jkotas commented on src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp:
```diff
@@ -430,9 +459,14 @@ REDHAWK_PALEXPORT bool REDHAWK_PALAPI PalInit()
     InitializeCurrentProcessCpuCount();
 
+#ifdef TARGET_LINUX
+    pthread_once(&once, CreateKey);
```
`PalInit` is called only once. `pthread_once` is unnecessary

jkotas commented on src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp:
```diff
@@ -481,10 +512,19 @@ EXTERN_C intptr_t RhGetCurrentThunkContext()
-    tls_destructionMonitor.SetThread(thread);
+#ifdef TARGET_LINUX
+    Link* link;
+    link = (Link*)malloc(sizeof(Link));
```
`PalAttachThread` is only called once per thread. The linked list is unnecessary

am11 replied:
Without this linked list, linux builds were failing in previous attempts. I spent some time to figure this one out by experimenting in linux-x64 and linux-musl-x64. I'm not sure how to make it work without a linked list.

jkotas replied:
I have left suggestion with I think it should look like.

jkotas commented on src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp:
```diff
@@ -406,6 +402,39 @@ void InitializeCurrentProcessCpuCount()
+
+static void CreateKey(void)
+{
+    pthread_key_create(&key, CallDestructors);
```
```suggestion
    pthread_key_create(&key, RuntimeThreadShutdown);
```

am11 replied:
I have cleaned this up. This refactoring is working after moving the key creation to PalInit. When we were using `__cxa_thread_exit`, linked list was required then. Thanks for pushing it. :)

(waiting for CI to see how `-fno-threadsafe-statics` pans out, locally osx and linux were happy)

am11 commented:
@jkotas, I've addressed the feedback. Can you take another look?

>  Compiling cpp files with the expectation they aren't using the C++ runtime but on what is acceptable is rather annoying. 

@AaronRobinsonMSFT, the alternative is to convert C++ compiled for NativeAOT to C, which the other comments suggest that we do not want to do at this point. Any other solution would be as good or as bad as this one.

jkotas commented on src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp:
```diff
@@ -430,11 +430,14 @@ REDHAWK_PALEXPORT bool REDHAWK_PALAPI PalInit()
     InitializeCurrentProcessCpuCount();
 
+#ifdef TARGET_LINUX
+    pthread_key_create(&key, RuntimeThreadShutdown);
```
error handling?

jkotas commented on src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp:
```diff
@@ -481,10 +483,10 @@ EXTERN_C intptr_t RhGetCurrentThunkContext()
-#if HAVE_THREAD_LOCAL
-    tls_destructionMonitor.SetThread(thread);
+#ifdef TARGET_LINUX
+    pthread_setspecific(key, thread);
```
Error handling? We probably just want to fail-fast when this fails. There is no good way to recover.

jkotas commented on src/coreclr/nativeaot/Runtime/stdcppshim.cpp:
```diff
@@ -0,0 +1,24 @@
+    return malloc(n);
+}
+
+void operator delete(void *p) noexcept
```
Can we do something about these so that they are not globally visible? Make them static?

am11 replied:
They are not visible outside the object. It is the same (global scoped) approach in `src/coreclr/utilcode/clrhost_nodependencies.cpp`.

The standard doesn't allow namespace scoped operator delete, it allows class/struct scoped one, but they cannot be used in global scope.

jkotas replied:
What makes it not visible outside the object? What would an implementation that is visible outside the object look like?

> It is the same (global scoped) approach in src/coreclr/utilcode/clrhost_nodependencies.cpp.

I know we had been fighting problems with this one, and I do not think we found good solution

am11 replied:
With `extern "C"`, it will have a visible symbol which will conflict with other object depending on libstdc++.

AaronRobinsonMSFT replied:
The symbol should still be visible though. There is nothing here that stops them from being visible during link. I could be missing something here, but they are defined in the header and not marked `static` so they can be discovered outside of the compilation unit. Is this some NativeAOT pattern I'm missing?

am11 replied:
> There is nothing here that stops them from being visible during link. 

There is this: https://github.com/dotnet/runtime/blob/e28cbaed0fdda578d7e365495e7821bfeb58e194/eng/native/configurecompiler.cmake#L482

jkotas replied:
This controls visibility in the .so after linking. I do not think that this controls visibility between multiple .o files.

AaronRobinsonMSFT replied:
Perhaps we aren't talking about the same thing or I could be misunderstanding something with that flag. My reading of `-fvisibility` is it relates to post-link exports of a dynamic library (`.so`) not the object files creating during compilation and then assembled at link-time. That is at least how `__declspec(dllexport)` works. It doesn't hide them from other compilation units during link, rather it presents them as exports from the linked library (that is, `.dll`). Those functions will still be visible to other compilation units pre-link.

am11 replied:
I image there is no tests for the scenario we are worried will break, right? Is it important enough to block this PR on?

AaronRobinsonMSFT replied:
> I image there is no tests for the scenario we are worried will break, right? 

I would agree with that. @jkotas's concern is something I share and along with @janvorli have had to chase down - binding to incorrect allocators. This is likely far more manageable in this narrow slice of the world, but it has caused us a non-trivial amount of stress and fire drills, particularly in diagnostic scenarios.

am11 replied:
One option is that we create a [`gcompat`](https://git.adelielinux.org/adelie/gcompat/) type object with these few definitions and link it by default. For static NativeAOT publishing, we guide the user to drop it if they are using `-lstdc++`.

jkotas replied:
I would still keep most of the changes you have made in this PR, so that the compat lib can be minimal.

jkotas commented on src/coreclr/nativeaot/Runtime/inc/CommonTypes.h:
```diff
@@ -21,4 +43,6 @@ typedef uint32_t            UInt32_BOOL;    // windows 4-byte BOOL, 0 -> false,
 #define UInt32_FALSE        0
 #define UInt32_TRUE         1
 
+#define ABSTRACT_IMPL_NO_PURE_VIRTUAL { while(true); assert(!"abstract method"); }
```
The assert will never be hit - it is after the infinite while loop.

I am surprised that this compiles. I would expect that the compiler will complain about missing return value or unreachable code.

am11 replied:
Should we keep the loop and delete assert?

jkotas replied:
I think we want to have assert in case this gets called. I had the loop there just to stop compiler from complaining about missing return value.

jkotas commented on src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp:
```diff
@@ -430,11 +430,18 @@ REDHAWK_PALEXPORT bool REDHAWK_PALAPI PalInit()
+#ifdef TARGET_LINUX
+    if (pthread_key_create(&key, RuntimeThreadShutdown) != 0)
+    {
+        _ASSERTE(!"pthread_key_create failed");
+        RhFailFast();
```
```suggestion
        return false;
```

AaronRobinsonMSFT commented on src/coreclr/gc/unix/gcenv.unix.cpp:
```diff
@@ -38,7 +33,9 @@
 #endif
 
+#ifdef TARGET_WINDOWS
 #include <algorithm>
```
out of curiosity, what algorithms are we using?

am11 replied:
It was used for min/max which I have replaced (since it was colliding with std header). I just forgot to this clean it up (TARGET_WINDOWS is a nop in this file). :)

AaronRobinsonMSFT replied:
I'd also be fine creating a `MAX` or `MIN` macro to share. The lowercase variety are not acceptable but uppercase I can work with.

jkotas commented on src/coreclr/nativeaot/Runtime/inc/CommonTypes.h:
```diff
@@ -6,10 +6,35 @@
+#include <new>
 using std::nothrow;
+
+#define VIRTUAL_IMPL = 0;
```
Nit: I think "pure" is an important to describe what this does. I would call this `PURE_VIRTUAL_IMPL` or just `PURE_VIRTUAL`.

am11 replied:
The one in #else branch is not pure, so I went with a neutral name.

jkotas replied:
It acts like a pure virtual impl in both cases. I think the name should describe what it does, not how it is implemented.

jkotas commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:
```diff
@@ -48,7 +48,7 @@ The .NET Foundation licenses this file to you under the MIT license.
       <NativeLibrary Condition="$(NativeLib) == ''" Include="$(IlcSdkPath)libbootstrapper.a" />
       <NativeLibrary Condition="$(NativeLib) != ''" Include="$(IlcSdkPath)libbootstrapperdll.a" />
-      <NativeLibrary Include="$(IlcSdkPath)$(FullRuntimeName).a" />
+      <NativeLibrary Include="$(IlcSdkPath)$(FullRuntimeName).a;$(IlcSdkPath)libstdc++compat.a" />
```
Could you please add property for this, something like `LinkStandardCPlusPlusLibrary`. When it is not set, use the shim. When it is set, skip the shim and use `-lstdc++` instead.

Also, set this property in one of the smoke tests so that we know that it works.

jkotas replied:
Any particular reason for naming the property `SkipCppCompat`?

I prefer for the public surface to be explainable in terms of other public surface and not in terms of internal implementation details. "This enables linking with stdc++ library" is much easier to explain than "This disables internal implementation detail that may cause problems when linking with stdc++ library".

Also, could you please also set it in one of the smoke tests, for example here: https://github.com/dotnet/runtime/blob/main/src/tests/nativeaot/SmokeTests/Preinitialization/Preinitialization.csproj#L13 so that we have validation that it works?

am11 replied:
> Any particular reason for naming the property SkipCppCompat?

I was already working and testing on this infra before I saw your comment. I'm still working on linux, will push updates once linux is fixed.

am11 committed:
Remove libstdc++ dependency from NativeAOT

am11 committed:
Revert a few files

jkotas reviewed (approved):
Looks great to me! 

@AaronRobinsonMSFT Any other feedback?

cc @dotnet/gc for awareness. This contains some minor changes in GC/EE interface contract. These changes do not break binary compatibility - newer standalone gc builds should continue to work with older runtimes, just like before.

mangod9 commented on src/coreclr/gc/gcinterface.h:
```diff
@@ -485,64 +485,68 @@ typedef void (* fq_scan_fn)(Object** ppObject, ScanContext *pSC, uint32_t dwFlag
 
-
+#if defined(FEATURE_NATIVEAOT) && !defined(TARGET_WINDOWS)
+#define PURE_VIRTUAL { assert(!"pure virtual function called"); while(true); }
```
`while(true)` looks odd here, is there any precedent where its done for similar cases?

jkotas replied:
It can use a comment on what it is trying to achieve.

We have a similar infinite loop for example here: https://github.com/dotnet/runtime/blob/9d6396deb02161f5ee47af72ccac52c2e1bae458/src/coreclr/vm/fcall.h#L814 (it is done for a different purpose, but it would be very confusing without the comment).

AaronRobinsonMSFT replied:
This is an infinite spin so we can find issues faster in release builds. I agree this could be annoying. I'm fine with it as is, but perhaps we could instead insert a `FAILFAST`? @jkotas or @am11 Any push back?

jkotas replied:
I would keep this as simple as possible. The infinite loop is here since we do not have a good value to return. The chance that somebody will hit it in release build without seeing the asset in debug builds before are close to zero.

Also, it won't be an infinite loop in release build, when compiled with recent clang at least. It will get optimized into nothing (yes, C/C++ standard allows that).

am11 replied:
> It will get optimized into nothing (yes, C/C++ standard allows that)

Indeed https://github.com/llvm/llvm-project/blob/f5cbbb998826bc84cbebd12f890622b3ec3184fe/llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp#L1018-L1021 😅

mangod9 replied:
a comment would certainly help to reduce confusion for the reader here :)

am11 commented on src/coreclr/gc/gcinterface.h:
```diff
@@ -485,64 +485,68 @@ typedef void (* fq_scan_fn)(Object** ppObject, ScanContext *pSC, uint32_t dwFlag
 typedef bool (* async_pin_enum_fn)(Object* object, void* context);
 
-
+#if defined(FEATURE_NATIVEAOT) && !defined(TARGET_WINDOWS)
```
```suggestion
// Implement pure virtual on NativeAOT Unix (for -p:LinkStandardCPlusPlusLibrary=false the default),
// to avoid linker requiring __cxa_pure_virtual.
#if defined(FEATURE_NATIVEAOT) && !defined(TARGET_WINDOWS)
// `while(true);` is to satisfy the missing `return` statement. It will be optimized away by the compiler.
```

am11 commented on src/coreclr/nativeaot/Runtime/inc/CommonTypes.h:
```diff
@@ -6,9 +6,16 @@
+#include <stdio.h>
 #include <new>
 
+#ifdef TARGET_WINDOWS
+#define PURE_VIRTUAL = 0;
+#else
+#define PURE_VIRTUAL { assert(!"pure virtual function called"); while(true); }
```
```suggestion
// Implement pure virtual on NativeAOT Unix (for -p:LinkStandardCPlusPlusLibrary=false the default),
// to avoid linker requiring __cxa_pure_virtual.
#ifdef TARGET_WINDOWS
#define PURE_VIRTUAL = 0;
#else
// `while(true);` is to satisfy the missing `return` statement. It will be optimized away by the compiler.
#define PURE_VIRTUAL { assert(!"pure virtual function called"); while(true); }
```

am11 commented on src/tests/nativeaot/SmokeTests/Preinitialization/Preinitialization.csproj:
```diff
@@ -11,6 +11,12 @@
     <IlcTrimMetadata>false</IlcTrimMetadata>
+
+    <!--
+      By default, we link libstdc++compat.a, this property is set to exercise code
```
```suggestion
      By default, we link libstdc++compat.a, This property is set to exercise code
```

am11 committed:
Add comments

mangod9 reviewed (approved):
thx!

jkotas commented:
Failure is https://github.com/dotnet/arcade/issues/11239

am11 commented:
@EgorBo, crossgen2 binary size with final changes:

macOS: +272 bytes
Linux: -10608 bytes
